skbio/io/format/fasta.py

```diff
@@ -940,7 +948,7 @@ def _parse_fasta_raw(fh, data_parser, error_type):
                 data_chunks.append(line)
         prev = line
     # yield last record in file
-    yield data_parser(data_chunks), id_, desc
+    yield data_parser(data_chunks), id_, desc, consumed
```

Comment:
I'm worried about separation of concerns here. The parser is responsible for reporting to the callee about what data it consumed, but the callee already knows what data it is providing. This adds unnecessary bookkeeping to the parser when this could be managed higher in the stack

Reply:
Hey Daniel! Me and @qiyunzhu discussed this, and I wanted to clarify that every file format has a unique parser within the scikit-bio system. Due to this, every sniffer reads a different amount of data at a different time, which is hard to predict before getting to this sniffer. That's why I decided to do parsing at the sniffer level instead of higher up. Doing it higher up also runs the risk of incorrectly predicting how much to give to the sniffer, resulting in false negatives.

Reply:
Re: @wasade I discussed with Raeed and am under impression that the current solution is a necessary evil. The entire I/O may be revisited, and Matt may be able to work on it. But at this moment, I can't think of a better way to handle stdin.

Reply:
We shouldn't expect the sniffers to be perfect. There are a lot of pathological cases for the common formats here. We don't know _for sure_ that we can parse until we actually fully parse, but that is expensive which is why we sniff.  

Why not standardize on reading and sniffing the first 1024 bytes -- are there any sniffers which actually need more than that?

Reply:
@qiyunzhu, something like [this](https://github.com/wasade/mxdx/blob/main/mxdx/_io.py#L285-L301) doesn't work?

Reply:
@wasade I am afraid that some sniffers need more than that, and it is hard to predict which one unless we examine every existing sniffer. It could also be that some sniffers (like csv-based ones) assumes complete table structure and could break if the file content is truncated to the first 1024 bytes. I agree that your solution is ideal, and I prefer that too. But to adopt it, we need to investigate every single existing sniffer in scikit-bio and propose a revision plan. This is probably beyond Raeed's capacity, and should be reassigned to Matt.